sympy/series/sequences.py

```diff
@@ -0,0 +1,522 @@
+
+
+class SeqPer(SeqExpr):
+    """Represents a periodical sequence
+
+    The elements are repeated after a given period.
+
+    Examples
+    ========
+
+    >>> from sympy import SeqPer, oo
+    >>> s = SeqPer((1, 2, 3), (0, 5))
+    >>> s.periodical
+    (1, 2, 3)
+    >>> s.period
+    3
+
+    For value at a particular point
+
+    >>> s.coeff(3)
+    1
+
+    supports slicing
+
+    >>> s[:]
+    [1, 2, 3, 1, 2, 3]
+
+    iterable
+
+    >>> list(s)
+    [1, 2, 3, 1, 2, 3]
+
+    changing step size
+
+    >>> SeqPer((1, 2, 3), (0, 5, 2))[:]
+    [1, 3, 2]
+
+    sequence starts from negative infinity
+
+    >>> SeqPer((1, 2, 3), (-oo, 0))[0:6]
+    [1, 2, 3, 1, 2, 3]
+
+    See Also
+    ========
+
+    sympy.series.sequences.SeqFormula
+    sympy.series.sequences.SeqFunc
+    """
+
+    @property
+    def period(self):
+        return len(self.gen)
+
+    @property
+    def periodical(self):
+        return self.gen
+
+    def _eval_coeff(self, i):
+        if self.start is S.NegativeInfinity:
+            idx = (self.stop - i) % self.period
+        else:
+            idx = (i - self.start) % self.period
+        return self.periodical[idx]
+
+
+class SeqFormula(SeqExpr):
+    """Represents sequence based on a formula
+
+    Elements are generated using a formula
+
+    Examples
+    ========
+
+    >>> from sympy import SeqFormula, oo, Symbol
+    >>> n = Symbol('n')
+    >>> s = SeqFormula((n**2, n), (0, 5))
+    >>> s.formula
+    n**2
+
+    For value at a particular point
+
+    >>> s.coeff(3)
+    9
+
+    supports slicing
+
+    >>> s[:]
+    [0, 1, 4, 9, 16, 25]
+
+    iterable
+
+    >>> list(s)
+    [0, 1, 4, 9, 16, 25]
+
+    changing step size
+
+    >>> SeqFormula((n**2, n), (0, 5, 2))[:]
+    [0, 4, 16]
+
+    sequence starts from negative infinity
+
+    >>> SeqFormula((n**2, n), (-oo, 0))[0:6]
+    [0, 1, 4, 9, 16, 25]
+
+    See Also
+    ========
+
+    sympy.series.sequences.SeqPer
+    sympy.series.sequences.SeqFunc
+    """
+
+    def __new__(cls, formula, interval=(0, S.Infinity, 1)):
+        # try to find the dummy symbol
+        formula = sympify(formula)
+        if not is_sequence(formula, Tuple):
+            free = formula.free_symbols
+            if len(free) != 1:
+                raise ValueError(filldedent(
+                    " specify dummy variables for %s. If the formula contains"
+                    " more than one free symbol, a dummy variable should be"
+                    " supplied explicitly e.g., SeqFormula((m*n**2, n), (0, 5))"
+                    % formula))
+            formula = (formula, free.pop())
+        return SeqExpr.__new__(cls, formula, interval)
+
+    @property
+    def formula(self):
+        return self.gen[0]
+
+    @property
+    def _dummy(self):
+        return (self.gen[1],)
+
+    def _eval_coeff(self, i):
+        d = self.dummy[0]
+        return self.formula.subs(d, i)
+
+
+class SeqFunc(SeqExpr):
+    """Represents sequence based on a function
```

Comment:
Since normal callables can't be sympified, perhaps any callable isn't the best idea, and only `Lambda` function should be used. Some more thought needs to be put into this.